Doc/library/unittest.mock.rst

```diff
@@ -1299,13 +1299,21 @@ patch.dict
     decorator. When used as a class decorator :func:`patch.dict` honours
     ``patch.TEST_PREFIX`` for choosing which methods to wrap.
 
+.. versionchanged:: 3.7
+
+    :func:`patch.dict` now returns the patched dictionary when used as a context
+    manager.
+
 :func:`patch.dict` can be used to add members to a dictionary, or simply let a test
 change a dictionary, and ensure the dictionary is restored when the test
 ends.
 
     >>> foo = {}
-    >>> with patch.dict(foo, {'newkey': 'newvalue'}):
+    >>> with patch.dict(foo, {'newkey': 'newvalue'}) as patched_foo:
     ...     assert foo == {'newkey': 'newvalue'}
+    ...     assert patched_foo == {'newkey': 'newvalue'}
+    ...     patched_foo['spam'] = 'eggs'
+    ...     assert patched_foo['spam'] == 'eggs'
     ...
     >>> assert foo == {}
```

Comment:
Can you add example with patched_foo here to make behaviour clear to readers?

Reply:
Not sure I understand what you mean. `patched_foo` is to be used within the context manager.

Reply:
But it still exists after we exit the context manager block!

Reply:
Ah, since the patched dictionary is now returned by the context manager as `patched_foo`. The variable `patched_foo` will still be accessible outside and will be unpatched to have value of `{}` . So `assert foo == patched_foo` is a valid statement outside the context manager.

```python
from unittest.mock import patch

foo = {}

with patch.dict(foo, {'newkey': 'newvalue'}) as patched_foo:
    assert foo == {'newkey': 'newvalue'}
    assert patched_foo == {'newkey': 'newvalue'}
    patched_foo['spam'] = 'eggs'
    assert patched_foo['spam'] == 'eggs'

assert foo == patched_foo
assert foo == {}
assert patched_foo == {}
```

Reply:
Yeah but that is not something you'd expect in the docs right? I can add a test about it in the test area but it feels like just adding noise here.

The same way you would not document open by using the file object after the context manager.

Reply:
Agreed that it would be little confusing in docs here since it's not expected that someone wants to use the context manager variable outside the block. It would also require a comment if added to explain `patched_foo` is unpatched though patched value was available inside the block and might require some understanding of the internals over how dict is patched and unpatched.

Reply:
I think file object is different because `__enter__` returns self, and it is uncommon to have two variables for the file object, like in this contrived example:

```
file1 = open("...")
with file1 as file2:
    data = file2.read()

assert file1.closed
assert file2.closed
```

The last two lines mean the same thing here, but I feel that for patch.dict it could be useful to add one line of doc to make it clear what people can or can’t do with `patched_dict` after the context block (i.e. is it the same dict object as `foo`, or a different object with the same contents, or the same contents as inside the context block).

Reply:
Added the extra check.